src/classes/CW_AudienceGroupControllerTest.cls

```diff
@@ -0,0 +1,333 @@
+			
+			agCtrl.cancel();
+			
+			system.assertEquals(agCtrl.inputMode,false);
+			
+			agCtrl.editGroups();
+			
+			agCtrl.getAudienceGroup();
+			
+			agCtrl.campGroups[0].aGroup.Description__c = 'Update Description';
+			agCtrl.campGroups[1].aGroup.Name = 'Update Name';
+			agCtrl.saveGroups();
+			
+			listGroups = [SELECT Id,Name,Description__c
+						  FROM CW_Audience_Group__c
+						  WHERE Campaign__c = : camp.Id];
+						  
+			system.assertEquals(listGroups[0].Description__c,'Update Description');
+			system.assertEquals(listGroups[1].Name,'Update Name');
+			
+			system.assertEquals(0,agCtrl.campGroups[0].getNumberOfMembers());
+			system.assertEquals(false,agCtrl.campGroups[0].getIsUsed());
+			
+			agCtrl.editGroups();
+			
+			agCtrl.getAudienceGroup();
+			
+			agCtrl.campGroups[1].aGroup.Name = 'Update Name';
+			agCtrl.saveGroups();
+			
+			for(ApexPages.Message msg :  ApexPages.getMessages()) 
+	    	{
+	    		system.assertEquals(msg.getSummary(),agCtrl.campGroups[1].aGroup.Name + ConstantUtil.DUPLICATE_AUDIENCE_GROUP_NAME);
+			}
+			
+			
+	    }
+	    Test.stopTest();
+    }
+    
+    static testMethod void downloadCSVFiles() 
+    {
+        initCustomSettings();
+        
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        insert camp;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,5);
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
+    	
+    	Test.startTest();
+    	
+		System.runAs(thisUser)
+	    {  
+	        Test.setCurrentPage(Page.CW_Campaign_Information);
+			CW_WizardController wizCtrl = new CW_WizardController(new ApexPages.StandardController(camp));
+			wizCtrl.next();
+			
+			Test.setCurrentPage(Page.CW_Audience);
+			Apexpages.currentPage().getParameters().put('id',camp.Id);
+			CW_AudienceGroupController agCtrl = new CW_AudienceGroupController();
+			
+			PageReference ref = new PageReference('/apex/CW_Audience_Group_Template');
+			ref.getParameters().put('id',camp.Id);
+			agCtrl.downloadAudienceGroupTemplate();
+			
+			ref = new PageReference('/apex/CW_Audience_Group_Members_Template');
+			ref.getParameters().put('id',camp.Id);
+			agCtrl.downloadGroupMembersTemplate();
+	    }
+	    Test.stopTest();
+    }
+    
+    static testMethod void changeTargetGroupToControlGroup() 
+    {
+        initCustomSettings();
+        
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        insert camp;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,5);
+        listGroups[0].Is_Control_Group__c = false;
+        listGroups[0].Control_Group__c = listGroups[1].Id;
+        listGroups[2].Is_Control_Group__c = false;
+        listGroups[2].Control_Group__c = listGroups[1].Id;
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.